File: src/main/java/com/anod/appwatcher/accounts/AuthTokenProvider.java
package com.anod.appwatcher.accounts;

import android.accounts.Account;
import android.accounts.AccountManager;
import android.accounts.AccountManagerFuture;
import android.accounts.AuthenticatorException;
import android.accounts.OperationCanceledException;
import android.app.Activity;
import android.content.Context;
import android.os.AsyncTask;
import android.os.Bundle;

import java.io.IOException;

import info.anodsplace.android.log.AppLog;

public class AuthTokenProvider {

    private static final String AUTH_TOKEN_TYPE = "androidmarket";
    public static final String ACCOUNT_TYPE = "com.google";
    private final AccountManager mAccountManager;

    public AuthTokenProvider(Context context) {
        mAccountManager = AccountManager.get(context);
    }


    public String requestTokenBlocking(Activity activity, Account acc) throws AuthenticatorException, OperationCanceledException, IOException {
        String token = getAuthToken(activity, acc);

        if (token != null) {
            mAccountManager.invalidateAuthToken(ACCOUNT_TYPE, token);
        }

        token = getAuthToken(activity, acc);
        return token;
    }

    public void requestToken(Activity activity, Account acc, AuthenticateCallback callback) {

        (new GetTokenTask(activity, acc, this, callback)).execute(0);

    }

    private String getAuthToken(Activity activity, Account acc) throws AuthenticatorException, OperationCanceledException, IOException {
        if (acc == null) {
            return null;
        }
        AccountManagerFuture<Bundle> future;

        future = mAccountManager.getAuthToken(
                acc, AUTH_TOKEN_TYPE, null, activity, null, null
        );

        String authToken = future.getResult().getString(AccountManager.KEY_AUTHTOKEN);
        return authToken;
    }

    public interface AuthenticateCallback {
        void onAuthTokenAvailable(String token);

        void onUnRecoverableException(final String errorMessage);
    }

    private static class GetTokenTask extends AsyncTask<Integer, Void, String> {
        private final AuthTokenProvider mAuthTokenProvider;
        private Account mAccount;
        private Activity mActivity;
        private AuthenticateCallback mCallback;

        public GetTokenTask(Activity activity, Account acc, AuthTokenProvider helper, AuthenticateCallback callback) {
            mAccount = acc;
            mActivity = activity;
            mCallback = callback;
            mAuthTokenProvider = helper;
        }

        @Override
        protected String doInBackground(Integer... params) {
            String token = null;
            try {
                token = mAuthTokenProvider.requestTokenBlocking(mActivity, mAccount);
            } catch (IOException | OperationCanceledException | AuthenticatorException e) {
                AppLog.e(e.getMessage(), e);
            }
            return token;
        }


        @Override
        protected void onPostExecute(String token) {
            super.onPostExecute(token);
            if (token == null) {
                mCallback.onUnRecoverableException("Cannot retrieve authorization token");
            } else {
                mCallback.onAuthTokenAvailable(token);
            }
        }
    }

}

File: src/main/java/com/anod/appwatcher/accounts/AuthTokenProvider.kt
package com.anod.appwatcher.accounts

import android.accounts.*
import android.accounts.AccountManager
import android.app.Activity
import android.content.Context
import android.os.AsyncTask
import android.os.Bundle
import com.anod.appwatcher.utils.BackgroundTask
import info.anodsplace.android.log.AppLog
import java.io.IOException

class AuthTokenProvider(context: Context) {
    private val accountManager: AccountManager = AccountManager.get(context)

    fun requestTokenBlocking(activity: Activity?, acc: Account): String {
        var token = ""
        try {
            token = getAuthToken(activity, acc)
        } catch (e: Exception) {
            AppLog.e(e)
        }

        if (token.isNotEmpty()) {
            accountManager.invalidateAuthToken(ACCOUNT_TYPE, token)
        }

        try {
            token = getAuthToken(activity, acc)
        } catch (e: Exception) {
            AppLog.e(e)
        }
        return token
    }

    fun requestToken(activity: Activity?, account: Account, callback: AuthenticateCallback) {
        BackgroundTask(object : BackgroundTask.Worker<Void?, String>(null) {
            override fun run(param: Void?): String {
                return requestTokenBlocking(activity, account)
            }

            override fun finished(result: String) {
                if (result.isEmpty()) {
                    callback.onUnRecoverableException("Cannot retrieve authorization token")
                } else {
                    callback.onAuthTokenAvailable(result)
                }
            }
        }).execute()
    }

    @Throws(AuthenticatorException::class, OperationCanceledException::class, IOException::class)
    private fun getAuthToken(activity: Activity?, acc: Account?): String {
        if (acc == null) {
            return ""
        }

        if (activity == null)
        {
            val future = accountManager.getAuthToken(acc, AUTH_TOKEN_TYPE, null, false, null, null)
            return future.result.getString(AccountManager.KEY_AUTHTOKEN) ?: ""
        }
        val future: AccountManagerFuture<Bundle> = accountManager.getAuthToken(
                acc, AUTH_TOKEN_TYPE, null, activity, null, null
        )
        return future.result.getString(AccountManager.KEY_AUTHTOKEN) ?: ""
    }

    interface AuthenticateCallback {
        fun onAuthTokenAvailable(token: String)
        fun onUnRecoverableException(errorMessage: String)
    }

    companion object {
        private const val AUTH_TOKEN_TYPE = "androidmarket"
        const val ACCOUNT_TYPE = "com.google"
    }
}
